docs/ci_send.3
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\" SPDX-License-Identifier: MIT
.\"
.TH CI_SEND 3 "25 July 1998"
.SH NAME
ci_send \- Initiate a DNS query
.SH SYNOPSIS
.nf
#include <ci.h>

typedef void (*ci_callback_dnsrec)(void *arg, ci_status_t status,
                                     size_t timeouts,
                                     const ci_dns_record_t *dnsrec);

ci_status_t ci_send_dnsrec(ci_channel_t *channel,
                               const ci_dns_record_t *dnsrec,
                               ci_callback_dnsrec callback,
                               void *arg, unsigned short *qid);

typedef void (*ci_callback)(void *arg, int status,
                              int timeouts, unsigned char *abuf,
                              int alen);

void ci_send(ci_channel_t *channel, const unsigned char *qbuf,
               int qlen, ci_callback callback, void *arg);

.fi
.SH DESCRIPTION
The \fIci_send_dnsrec(3)\fP function initiates a DNS query formatted using the
\fIci_dns_record_t *\fP data structure created via
\fIci_dns_record_create(3)\fP in the
.IR dnsrec
parameter.  The supplied callback in the
.IR callback
parameter also returns the response using a
\fIci_dns_record_t *\fP data structure.

The \fIci_send(3)\fP function similarly initiates a DNS query, but instead uses
raw binary buffers with fully formatted DNS messages passed in the request via the
.IR qbuf
and
.IR qlen
parameters. The supplied callback in the
.IR callback
parameter also returns the raw binary DNS response in the
.IR abuf
and
.IR alen
parameters. This method should be considered deprecated in favor of
\fIci_send_dnsrec(3)\fP.

Both functions take an initialized ci channel identified by
.IR channel .

The \fIci_send_dnsrec(3)\fP also can be supplied an optional output parameter of
.IR qid
to populate the query id as it was placed on the wire.

The \fIci_send_dnsrec(3)\fP function returns an \fIci_status_t\fP response
code.  This may be useful to know that the query was enqueued properly.  The
response code does not reflect the result of the query, just the result of the
enqueuing of the query.

Completion or failure of the query may happen immediately (even before the
function returning), or may happen later as network events are processed.

When the associated callback is called, it is called with a channel lock so care
must be taken to ensure any processing is minimal to prevent DNS channel stalls.

The callback may be triggered from a different thread than the one which
called \fIci_send_dnsrec(3)\fP or \fIci_send(3)\fP.

For integrators running their own event loops and not using \fBCI_OPT_EVENT_THREAD\fP,
care needs to be taken to ensure any file descriptor lists are updated immediately
within the eventloop when notified.

The callback argument
.IR arg
is copied from the \fIci_send_dnsrec(3)\fP or \fIci_send(3)\fP
.IR arg
parameter.

The callback argument
.I status
indicates whether the query succeeded and, if not, how it failed.  It
may have any of the following values:
.TP 19
.B CI_SUCCESS
The query completed.
.TP 19
.B CI_EBADQUERY
The query buffer was poorly formed (was not long enough for a DNS
header or was too long for TCP transmission).
.TP 19
.B CI_ETIMEOUT
No name servers responded within the timeout period.
.TP 19
.B CI_ECONNREFUSED
No name servers could be contacted.
.TP 19
.B CI_ENOMEM
Memory was exhausted.
.TP 19
.B CI_ECANCELLED
The query was cancelled.
.TP 19
.B CI_EDESTRUCTION
The name service channel
.I channel
is being destroyed; the query will not be completed.
.TP 19
.B CI_ENOSERVER
The query will not be completed because no DNS servers were configured on the
channel.
.PP

The callback argument
.I timeouts
reports how many times a query timed out during the execution of the
given request.

If the query completed, the callback argument
.IR dnsrec
for \fIci_send_dnsrec(3)\fP or
.IR abuf
and
.IR alen
for \fIci_send(3)\fP will be non-NULL.

Unless the flag
.B CI_FLAG_NOCHECKRESP
was set at channel initialization time, \fIci_send_dnsrec(3)\fP and
\fIci_send(3)\fP will normally ignore responses whose questions do not match
the supplied questions, as well as responses with reply codes of
.BR SERVFAIL ,
.BR NOTIMP ,
and
.BR REFUSED .
Unlike other query functions in the ci library, however,
\fIci_send_dnsrec(3)\fP and \fIci_send(3)\fP do not inspect the header of
the reply packet to determine the error status, so a callback status of
.B CI_SUCCESS
does not reflect as much about the response as for other query functions.

.SH AVAILABILITY
\fBci_send_dnsrec(3)\fP was introduced in c-ci 1.28.0.

.SH SEE ALSO
.BR ci_dns_record_create (3),
.BR ci_process (3),
.BR ci_search (3),
.BR ci_dns_record (3)


RELEASE-NOTES.md
## c-ci version 1.34.2 - October 15 2024

This release contains a fix for downstream packages detecting the c-ci
version based on the contents of the header file rather than the
distributed pkgconf or cmake files.

## c-ci version 1.34.1 - October 9 2024

This release fixes a packaging issue.


## c-ci version 1.34.0 - October 9 2024

This is a feature and bugfix release.

Features:
* adig: read arguments from adigrc.
  [PR #856](https://github.com/c-ci/c-ci/pull/856)
* Add new pending write callback optimization via `ci_set_pending_write_cb`.
  [PR #857](https://github.com/c-ci/c-ci/pull/857)
* New function `ci_process_fds()`.
  [PR #875](https://github.com/c-ci/c-ci/pull/875)
* Failed servers should be probed rather than redirecting queries which could
  cause unexpected latency.
  [PR #877](https://github.com/c-ci/c-ci/pull/877)
* adig: rework command line arguments to mimic dig from bind.
  [PR #890](https://github.com/c-ci/c-ci/pull/890)
* Add new method for overriding network functions
  `ci_set_socket_function_ex()` to properly support all new functionality.
  [PR #894](https://github.com/c-ci/c-ci/pull/894)
* Fix regression with custom socket callbacks due to DNS cookie support.
  [PR #895](https://github.com/c-ci/c-ci/pull/895)
* ci_socket: set IP_BIND_ADDRESS_NO_PORT on ci_set_local_ip* tcp sockets
  [PR #887](https://github.com/c-ci/c-ci/pull/887)
* URI parser/writer for ci_set_servers_csv()/ci_get_servers_csv().
  [PR #882](https://github.com/c-ci/c-ci/pull/882)

Changes:
* Connection handling modularization.
  [PR #857](https://github.com/c-ci/c-ci/pull/857),
  [PR #876](https://github.com/c-ci/c-ci/pull/876)
* Expose library/utility functions to tools.
  [PR #860](https://github.com/c-ci/c-ci/pull/860)
* Remove `ci__` prefix, just use `ci_` for internal functions.
  [PR #872](https://github.com/c-ci/c-ci/pull/872)


Bugfixes:
* fix: potential WIN32_LEAN_AND_MEAN redefinition.
  [PR #869](https://github.com/c-ci/c-ci/pull/869)
* Fix googletest v1.15 compatibility.
  [PR #874](https://github.com/c-ci/c-ci/pull/874)
* Fix pkgconfig thread dependencies.
  [PR #884](https://github.com/c-ci/c-ci/pull/884)


Thanks go to these friendly people for their efforts and contributions for this
release:

* Brad House (@bradh352)
* Cristian Rodríguez (@crrodriguez)
* Georg (@tacerus)
* @lifenjoiner
* Shelley Vohr (@codebytere)
* 前进，前进，进 (@leleliu008)



src/lib/ci_send.c
/* MIT License
 *
 * Copyright (c) 1998 Massachusetts Institute of Technology
 * Copyright (c) The c-ci project and its contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 * SPDX-License-Identifier: MIT
 */

#include "ci_private.h"

#ifdef HAVE_NETINET_IN_H
#  include <netinet/in.h>
#endif
#include "ci_nameser.h"

static unsigned short generate_unique_qid(ci_channel_t *channel)
{
  unsigned short id;

  do {
    id = ci_generate_new_id(channel->rand_state);
  } while (ci_htable_szvp_get(channel->queries_by_qid, id, NULL));

  return id;
}

/* https://datatracker.ietf.org/doc/html/draft-vixie-dnsext-dns0x20-00 */
static ci_status_t ci_apply_dns0x20(ci_channel_t    *channel,
                                        ci_dns_record_t *dnsrec)
{
  ci_status_t status = CI_SUCCESS;
  const char   *name   = NULL;
  char          dns0x20name[256];
  unsigned char randdata[256 / 8];
  size_t        len;
  size_t        remaining_bits;
  size_t        total_bits;
  size_t        i;

  status = ci_dns_record_query_get(dnsrec, 0, &name, NULL, NULL);
  if (status != CI_SUCCESS) {
    goto done;
  }

  len = ci_strlen(name);
  if (len == 0) {
    return CI_SUCCESS;
  }

  if (len >= sizeof(dns0x20name)) {
    status = CI_EBADNAME;
    goto done;
  }

  memset(dns0x20name, 0, sizeof(dns0x20name));

  /* Fetch the minimum amount of random data we'd need for the string, which
   * is 1 bit per byte */
  total_bits     = ((len + 7) / 8) * 8;
  remaining_bits = total_bits;
  ci_rand_bytes(channel->rand_state, randdata, total_bits / 8);

  /* Randomly apply 0x20 to name */
  for (i = 0; i < len; i++) {
    size_t bit;

    /* Only apply 0x20 to alpha characters */
    if (!ci_isalpha(name[i])) {
      dns0x20name[i] = name[i];
      continue;
    }

    /* coin flip */
    bit = total_bits - remaining_bits;
    if (randdata[bit / 8] & (1 << (bit % 8))) {
      dns0x20name[i] = name[i] | 0x20;                          /* Set 0x20 */
    } else {
      dns0x20name[i] = (char)(((unsigned char)name[i]) & 0xDF); /* Unset 0x20 */
    }
    remaining_bits--;
  }

  status = ci_dns_record_query_set_name(dnsrec, 0, dns0x20name);

done:
  return status;
}

ci_status_t ci_send_nolock(ci_channel_t *channel, ci_server_t *server,
                               ci_send_flags_t        flags,
                               const ci_dns_record_t *dnsrec,
                               ci_callback_dnsrec callback, void *arg,
                               unsigned short *qid)
{
  ci_query_t            *query;
  ci_timeval_t           now;
  ci_status_t            status;
  unsigned short           id          = generate_unique_qid(channel);
  const ci_dns_record_t *dnsrec_resp = NULL;

  ci_tvnow(&now);

  if (ci_slist_len(channel->servers) == 0) {
    callback(arg, CI_ENOSERVER, 0, NULL);
    return CI_ENOSERVER;
  }

  if (!(flags & CI_SEND_FLAG_NOCACHE)) {
    /* Check query cache */
    status = ci_qcache_fetch(channel, &now, dnsrec, &dnsrec_resp);
    if (status != CI_ENOTFOUND) {
      /* CI_SUCCESS means we retrieved the cache, anything else is a critical
       * failure, all result in termination */
      callback(arg, status, 0, dnsrec_resp);
      return status;
    }
  }

  /* Allocate space for query and allocated fields. */
  query = ci_malloc(sizeof(ci_query_t));
  if (!query) {
    callback(arg, CI_ENOMEM, 0, NULL); /* LCOV_EXCL_LINE: OutOfMemory */
    return CI_ENOMEM;                  /* LCOV_EXCL_LINE: OutOfMemory */
  }
  memset(query, 0, sizeof(*query));

  query->channel      = channel;
  query->qid          = id;
  query->timeout.sec  = 0;
  query->timeout.usec = 0;
  query->using_tcp =
    (channel->flags & CI_FLAG_USEVC) ? CI_TRUE : CI_FALSE;

  /* Duplicate Query */
  status = ci_dns_record_duplicate_ex(&query->query, dnsrec);
  if (status != CI_SUCCESS) {
    ci_free(query);
    callback(arg, status, 0, NULL);
    return status;
  }

  ci_dns_record_set_id(query->query, id);

  if (channel->flags & CI_FLAG_DNS0x20 && !query->using_tcp) {
    status = ci_apply_dns0x20(channel, query->query);
    if (status != CI_SUCCESS) {
      /* LCOV_EXCL_START: OutOfMemory */
      callback(arg, status, 0, NULL);
      ci_free_query(query);
      return status;
      /* LCOV_EXCL_STOP */
    }
  }

  /* Fill in query arguments. */
  query->callback = callback;
  query->arg      = arg;

  /* Initialize query status. */
  query->try_count = 0;

  if (flags & CI_SEND_FLAG_NORETRY) {
    query->no_retries = CI_TRUE;
  }

  query->error_status = CI_SUCCESS;
  query->timeouts     = 0;

  /* Initialize our list nodes. */
  query->node_queries_by_timeout = NULL;
  query->node_queries_to_conn    = NULL;

  /* Chain the query into the list of all queries. */
  query->node_all_queries = ci_llist_insert_last(channel->all_queries, query);
  if (query->node_all_queries == NULL) {
    /* LCOV_EXCL_START: OutOfMemory */
    callback(arg, CI_ENOMEM, 0, NULL);
    ci_free_query(query);
    return CI_ENOMEM;
    /* LCOV_EXCL_STOP */
  }

  /* Keep track of queries bucketed by qid, so we can process DNS
   * responses quickly.
   */
  if (!ci_htable_szvp_insert(channel->queries_by_qid, query->qid, query)) {
    /* LCOV_EXCL_START: OutOfMemory */
    callback(arg, CI_ENOMEM, 0, NULL);
    ci_free_query(query);
    return CI_ENOMEM;
    /* LCOV_EXCL_STOP */
  }

  /* Perform the first query action. */

  status = ci_send_query(server, query, &now);
  if (status == CI_SUCCESS && qid) {
    *qid = id;
  }
  return status;
}

ci_status_t ci_send_dnsrec(ci_channel_t          *channel,
                               const ci_dns_record_t *dnsrec,
                               ci_callback_dnsrec callback, void *arg,
                               unsigned short *qid)
{
  ci_status_t status;

  if (channel == NULL) {
    return CI_EFORMERR; /* LCOV_EXCL_LINE: DefensiveCoding */
  }

  ci_channel_lock(channel);

  status = ci_send_nolock(channel, NULL, 0, dnsrec, callback, arg, qid);

  ci_channel_unlock(channel);

  return status;
}

void ci_send(ci_channel_t *channel, const unsigned char *qbuf, int qlen,
               ci_callback callback, void *arg)
{
  ci_dns_record_t *dnsrec = NULL;
  ci_status_t      status;
  void              *carg = NULL;

  if (channel == NULL) {
    return;
  }

  /* Verify that the query is at least long enough to hold the header. */
  if (qlen < HFIXEDSZ || qlen >= (1 << 16)) {
    callback(arg, CI_EBADQUERY, 0, NULL, 0);
    return;
  }

  status = ci_dns_parse(qbuf, (size_t)qlen, 0, &dnsrec);
  if (status != CI_SUCCESS) {
    callback(arg, (int)status, 0, NULL, 0);
    return;
  }

  carg = ci_dnsrec_convert_arg(callback, arg);
  if (carg == NULL) {
    /* LCOV_EXCL_START: OutOfMemory */
    status = CI_ENOMEM;
    ci_dns_record_destroy(dnsrec);
    callback(arg, (int)status, 0, NULL, 0);
    return;
    /* LCOV_EXCL_STOP */
  }

  ci_send_dnsrec(channel, dnsrec, ci_dnsrec_convert_cb, carg, NULL);

  ci_dns_record_destroy(dnsrec);
}

size_t ci_queue_active_queries(const ci_channel_t *channel)
{
  size_t len;

  if (channel == NULL) {
    return 0;
  }

  ci_channel_lock(channel);

  len = ci_llist_len(channel->all_queries);

  ci_channel_unlock(channel);

  return len;
}
